test/beginend-dired-test.el

```diff
@@ -33,80 +33,110 @@
+
+(defmacro with-beginend-dired-test-buffer (specs entries &rest body)
+  "Execute the forms in BODY within a temporary directory.
+The directory is created by SPECS and ENTRIES.  See the documentation of
+`beginend-create-dired-buffer'."
+  (declare (indent 2))
+  `(with-current-buffer (beginend-create-dired-buffer ',specs ,entries)
+     ,@body))
 
 (describe "beginend in a dired buffer"
-  (describe "without dired-hide-details-mode"
-    (describe "without dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p))
-
-      (it "offers a command to go to last file"
-        (beginend-dired-test)))
-
-    (describe "with dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'omit))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'omit))))
-
-  (describe "with-dired-hide-details-mode (also hides information lines)"
-    (describe "without dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'hide-details))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'hide-details)))
-
-    (describe "with dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'omit 'hide-details))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'omit 'hide-details))))
-
-  (describe "with-dired-hide-details-mode (does not hide information lines)"
-    (describe "without dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'hide-details 'keep-info-lines))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'hide-details 'keep-info-lines)))
-
-    (describe "with dired-omit-mode"
-      (it "offers a command to go to first file"
-        (beginend-dired-test 'begin-p 'omit 'hide-details 'keep-info-lines))
-
-      (it "offers a command to go to last file with dired-omit-mode"
-        (beginend-dired-test 'omit 'hide-details 'keep-info-lines)))))
+  (it "tests . and .. at the beginning"
+    (with-beginend-dired-test-buffer (dotfiles)
+        '("." ".." "dir1" "dir2")
+      (beginend-dired-test-begin "dir1")
+      (beginend-dired-test-end "dir2")))
+  (it "tests . and .. at the end"
+    (with-beginend-dired-test-buffer (dotfiles)
+        '("dir1" "dir2" "." "..")
+      (beginend-dired-test-begin "dir1")
+      (beginend-dired-test-end "dir2")))
+  (it "tests . at the beginning and .. at the end"
+    (with-beginend-dired-test-buffer (dotfiles)
+        '("." "dir1" "dir2" "..")
+      (beginend-dired-test-begin "dir1")
+      (beginend-dired-test-end "dir2")))
+  (it "tests no . and .."
+    (with-beginend-dired-test-buffer ()
+        '("." ".." "dir1" "dir2")
```

Comment:
it's not clear to me why the test says "no . and .." but you still include both in the list What about

```emacs
(it "ignores . and .. when they are hidden"
    ...)
```